### PR TITLE
[BACK-513] Switching to FastAPI

### PR DESCRIPTION
**Pull request by bassrock**
## Goal

Convert the Flask portion of this app to Fast

## Todos:
- [x] Clean up docker-compose
- [x] Switch to Pipenv
- [x] Convert Flask calls to FastAPI [(super helpful doc)](https://amitness.com/2020/06/fastapi-vs-flask/)
- [x] Migrate all code into app
- [ ] Test on dev

## Implementation Decisions

I chose to not switch to any of the fancy FastAPI pydantic models for request bodies, do to not knowing the full nature of the request bodies, but I would like to come back and do another cleanup pass.  

I also did not go through everything to make it fully asyncio (I also didn;t see places where this would be an immediate benefit), I figured converting to FastAPI was a big enough change to start and we can come back and do more improvements.

There are also some things I'd like to do with the GeoMind DB and baking that into the docker image.

I also had to move all code into `app` so that it would work with gunicorn.

## All Submissions:

- [x] Have you followed the guidelines in our Contributing document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### CODE REVIEW

**Comment by bassrock** [comment]
@mmiermans How can I deploy this to the dev infrastructure?

File: tests/unit/test_app.py [hunk 1]
```diff
@@ -46,7 +42,7 @@ def get_request_body(cls, without=None, placements=None, update=None):
     @patch('provider.geo_provider.GeolocationProvider.__init__', return_value=None)
     def test_app_pulse(self, mock_geo):
         resp = self.create_client_no_geo_locs().get('/pulse')
-        self.assertEqual(resp.json, {"pulse" : "ok"})
+        self.assertEqual(resp.json(), {"pulse" : "ok"})
```
Comment (bassrock) [inline tests/unit/test_app.py:45]: Note that expect and actual are flipped in all these asserts. Will do another pr pass to flip these later.

Comment (chelseatroy) [reply tests/unit/test_app.py:45]: 👍

**Review by chelseatroy — APPROVED** [review]
LGTM, pending the switch on expected and actual!

**Review by kkyeboah — DISMISSED** [review]
Looks good to me pending tests on dev

**Comment by bassrock** [comment]
Based on the initial load tests, the change to FastAPI can actually handle the same number of requests the production service has on close to if not less than the same number of tasks.


I manually scaled the service to 100 tasks and it hit the same load production has today before some tasks instance started having trouble. It's also important to note that response times also stayed the same as Production today. 

It's also important to note the issues encountered was Nginx worker_connection limits which caused the  tasks to fail.


**Production Today**
Screen Shot 2021-01-21 at 4 39 47 PM


**Load Test**
Screen Shot 2021-01-21 at 4 38 52 PM

**Comment by bassrock** [comment]
Putting back to draft. @mmiermans and I decided we will deploy https://github.com/Pocket/proxy-server/pull/45 first and then we will need to rebase this to support it.

**Review by jpetto — COMMENTED** [review]
looks like we're good on reviews here. clearing from my queue.

**Review by mmiermans — APPROVED** [review]
- FastAPI load test: 7000 requests/second with 200 tasks @ 50% CPU
- Flask on production: 4350 requests/second with 200 tasks @ 44% CPU
- similar response times (P95 = 110-120ms, average = 64-67%)
- Production has a lower 5xx error rate at 0.1% than our load test at 0.5%, but that might be because our time-out configuration for aiohttp is slightly different that for requests.

Screenshot from 2021-01-25 17-15-09

**Review by mmiermans — APPROVED** [review]
Nice work! Great performance improvements, and some nice code and development improvements. I like moving all application code to the app directory.

There are some more code we can delete in app/main.py by leveraging FastAPI for parameter parsing, but you made the right call by leaving that for a future PR.

Nice work! :muscle:

**Pull request merged**